### PR TITLE
Feature/1 fix broken itsdangerous library

### PR DESCRIPTION
**Pull request by franciscoabenza**
After some struggles this is done.
the JSON Web Signature is replaced with a JWT + time expiration

### CODE REVIEW

**Comment by franciscoabenza** [comment]
Do not aprove .gitignore ( mistake found)

File: .gitignore [hunk 1]
```diff
@@ -4,6 +4,27 @@ tweet_data/
 data/
 venv/
 .vscode/
+<<<<<<< HEAD
```
Comment (zeyus) [inline .gitignore:7]: as mentioned in the comment this needs fixing, you can restore it to the original by doing `git checkout HEAD^1 -- .gitignore` (or something like that)

File: nlp4all/models.py [hunk 2]
```diff
@@ -322,18 +322,30 @@ class User(db.Model, UserMixin):
...
+            data = jwt.decode(
+                token,
+                app.config['SECRET_KEY'],
+                leeway=timedelta(seconds=10),
+                algorithms=["HS256"]
+                )
+
+            user_id = data['user_id']
+
+        except:
```
Comment (zeyus) [inline nlp4all/models.py:348]: this `except` is a bit too broad, are there any other exceptions expected? we can limit the scope by catching expected exceptions

File: nlp4all/routes.py [hunk 3]
```diff
@@ -587,7 +587,7 @@ def register_imc():
         hashed_password = bcrypt.generate_password_hash(
             fake_password).decode("utf-8")
         imc_org = Organization.query.filter_by(name="ATU").all()
-        a_project = imc_org[0].projects[0]
+        a_project = imc_org[0].projects[0] #error when no project. out of range TODO
```
Comment (zeyus) [inline nlp4all/routes.py:590]: good catch

**Review by zeyus — CHANGES_REQUESTED** [review]
Looks good, just a couple of mentioned changes :)

File: nlp4all/models.py [hunk 4]
```diff
@@ -322,18 +322,30 @@ class User(db.Model, UserMixin):
...
+                "exp": datetime.now(tz=timezone.utc)
+                        + timedelta(seconds=expires_sec)
+            },
+            app.config['SECRET_KEY'],
+            algorithm="HS256"
+        )
+        return reset_token
 
     @staticmethod
     def verify_reset_token(token):
```
Comment (zeyus) [inline nlp4all/models.py:337]: one other thing, the docstring has gone which makes linting fail, you could add something like

```python
"""Generate a password reset token

        Parameters:
            expires_sec (int): Number of seconds the token remains valid

        Returns:
            reset_token (str): The password reset token
"""
```

also as you've touched this function, feel free to add type hinting

e.g.

```python
def get_reset_token(self, expires_sec: int = 1800) -> str:
```

File: .gitignore (hunk 1 above)
Comment (franciscoabenza) [reply .gitignore:7]: I think this is changed now

File: nlp4all/models.py (hunk 2 above)
Comment (franciscoabenza) [reply nlp4all/models.py:348]: [only two reasonable errors to be expected(expired and invalid).
](https://pyjwt.readthedocs.io/en/stable/api.html#exceptions)
They are kinda handle where the function is called:
```python
 @app.route("/reset_password/<token>", methods=["GET", "POST"])

def reset_token(token):
    """Reset password token page"""
    if current_user.is_authenticated:
        return redirect(url_for("home"))
    user = User.verify_reset_token(token)
    if user is None:
        flash("That is an invalid or expired token", "warning")
        return redirect(url_for("reset_request"))
```

Tell me what you think :)

Comment (franciscoabenza) [reply nlp4all/models.py:348]: I could handle `InvalidSignatureError ` and `ExpiredSignatureError ` independently, but at the end of the day the user will have to try again.

File: nlp4all/routes.py (hunk 3 above)
Comment (franciscoabenza) [reply nlp4all/routes.py:590]: I think that this was back when running Python 3.10

File: nlp4all/models.py (hunk 4 above)
Comment (franciscoabenza) [reply nlp4all/models.py:337]: This should be done now :)

**Review by zeyus — APPROVED** [review]
Nice work!

**Pull request merged**